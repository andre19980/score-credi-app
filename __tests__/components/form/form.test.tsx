import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import StepForm from "@/components/form/step-form";

describe("Page", () => {
  it("renders form", () => {
    const { container } = render(<StepForm setUser={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });

  describe("personal info", () => {
    describe("input validation", () => {
      describe("`name` field", () => {
        it("is required", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });

          await userEvent.click(nextButton);

          expect(
            screen.getByText("O nome deve ter pelo menos 8 caracteres."),
          ).toBeInTheDocument();
        });

        it("accepts only non-numerical characters", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });
          const inputName = screen.getByLabelText("Nome");

          fireEvent.change(inputName, { target: { value: "12345678" } });
          await userEvent.click(nextButton);

          expect(
            screen.getByText("O nome não pode conter números."),
          ).toBeInTheDocument();
        });

        it("must be greater than 8 characters", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });
          const inputName = screen.getByLabelText("Nome");

          fireEvent.change(inputName, { target: { value: "name" } });
          await userEvent.click(nextButton);

          expect(
            screen.getByText("O nome deve ter pelo menos 8 caracteres."),
          ).toBeInTheDocument();
        });
      });

      describe("`age` field", () => {
        it("is required", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });

          await userEvent.click(nextButton);

          expect(
            screen.getByText("A idade deve ser maior ou igual a 18 anos."),
          ).toBeInTheDocument();
        });

        it("accepts only numerical characters", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });
          const inputAge = screen.getByLabelText("Idade");

          fireEvent.change(inputAge, { target: { value: "abc" } });
          await userEvent.click(nextButton);

          expect(
            screen.getByText("A idade deve ser maior ou igual a 18 anos."),
          ).toBeInTheDocument();
        });

        it("must be greater than 18", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });
          const inputAge = screen.getByLabelText("Idade");

          fireEvent.change(inputAge, { target: { value: "17" } });
          await userEvent.click(nextButton);

          expect(
            screen.getByText("A idade deve ser maior ou igual a 18 anos."),
          ).toBeInTheDocument();
        });

        it("must be less than 65", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });
          const inputAge = screen.getByLabelText("Idade");

          fireEvent.change(inputAge, { target: { value: "66" } });
          await userEvent.click(nextButton);

          expect(
            screen.getByText("A idade deve ser menor ou igual a 65 anos."),
          ).toBeInTheDocument();
        });
      });

      describe("`name` and `age`", () => {
        it("correct case", async () => {
          render(<StepForm setUser={jest.fn()} />);
          const nextButton = screen.getByRole("button", { name: "Próximo" });
          const inputName = screen.getByLabelText("Nome");
          const inputAge = screen.getByLabelText("Idade");

          fireEvent.change(inputAge, { target: { value: "30" } });
          fireEvent.change(inputName, { target: { value: "User Name Jr" } });

          await userEvent.click(nextButton);

          expect(screen.getByText("Cidade")).toBeInTheDocument();
        });
      });
    });
  });
});
