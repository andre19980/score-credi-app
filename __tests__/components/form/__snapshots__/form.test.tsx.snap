// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders form 1`] = `
<div>
  <form>
    <div
      style="opacity: 0; will-change: opacity;"
    >
      <h2
        class="text-base font-semibold leading-7 text-gray-900 mb-3"
      >
        Dados Pessoais
      </h2>
      <div
        class="grid grid-cols-1 gap-x-6 gap-y-4 sm:grid-cols-6 mt-4"
      >
        <div
          class="sm:col-span-4"
        >
          <label
            class="block text-sm font-medium leading-6 text-gray-900"
            for="name"
          >
            Nome
          </label>
          <div
            class="mt-2"
          >
            <div
              class="flex bg-white rounded-md shadow-sm ring-1 ring-inset ring-gray-300 focus-within:ring-2 focus-within:ring-inset focus-within:ring-indigo-600"
            >
              <input
                class="block flex-1 border-0 bg-transparent py-1.5 pl-1 text-gray-900 placeholder:text-gray-400 focus:ring-0 sm:text-sm sm:leading-6"
                id="name"
                name="name"
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          class="sm:col-span-2"
        >
          <label
            class="block text-sm font-medium leading-6 text-gray-900"
            for="age"
          >
            Idade
          </label>
          <div
            class="mt-2"
          >
            <div
              class="flex w-full bg-white rounded-md shadow-sm ring-1 ring-inset ring-gray-300 focus-within:ring-2 focus-within:ring-inset focus-within:ring-indigo-600"
            >
              <input
                class="block flex-1 w-full border-0 bg-transparent py-1.5 pl-1 text-gray-900 placeholder:text-gray-400 focus:ring-0 sm:text-sm sm:leading-6"
                id="age"
                name="age"
                type="number"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex justify-between pt-4 mt-8 border-t"
    >
      <button
        class="text-gray-600 bg-gray-100 hover:text-gray-600 py-2.5 px-5 text-sm font-medium focus:outline-none rounded-lg border border-gray-200 hover:bg-gray-100 focus:z-10 focus:ring-4 focus:ring-gray-100 dark:focus:ring-gray-700 dark:bg-gray-800 dark:text-gray-400 dark:border-gray-600 dark:hover:text-white dark:hover:bg-gray-700"
        disabled=""
        type="button"
      >
        Voltar
      </button>
      <button
        class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800"
        type="button"
      >
        Pr√≥ximo
      </button>
    </div>
  </form>
</div>
`;
